public/app/features/scenes/core/sceneGraph.ts

```diff
@@ -0,0 +1,104 @@
+import { getDefaultTimeRange, LoadingState } from '@grafana/data';
+
+import { sceneInterpolator } from '../variables/interpolation/sceneInterpolator';
+import { SceneVariableSet } from '../variables/sets/SceneVariableSet';
+import { SceneVariables } from '../variables/types';
+
+import { SceneDataNode } from './SceneDataNode';
+import { SceneTimeRange as SceneTimeRangeImpl } from './SceneTimeRange';
+import { SceneDataState, SceneEditor, SceneObject, SceneTimeRange } from './types';
+
+/**
+ * Get the closest node with variables
+ */
+export function getVariablesFor(sceneObject: SceneObject): SceneVariables {
+  if (sceneObject.state.$variables) {
+    return sceneObject.state.$variables;
+  }
+
+  if (sceneObject.parent) {
+    return getVariablesFor(sceneObject.parent);
+  }
+
+  return EmptyVariableSet;
+}
+
+/**
+ * Will walk up the scene object graph to the closest $data scene object
+ */
+export function getDataFor(sceneObject: SceneObject): SceneObject<SceneDataState> {
+  const { $data } = sceneObject.state;
+  if ($data) {
+    return $data;
+  }
+
+  if (sceneObject.parent) {
+    return getDataFor(sceneObject.parent);
+  }
+
+  return EmptyDataNode;
+}
+
+/**
+ * Will walk up the scene object graph to the closest $timeRange scene object
+ */
+export function getTimeRangeFor(sceneObject: SceneObject): SceneTimeRange {
+  const { $timeRange } = sceneObject.state;
+  if ($timeRange) {
+    return $timeRange;
+  }
+
+  if (sceneObject.parent) {
+    return getTimeRangeFor(sceneObject.parent);
+  }
+
+  return DefaultTimeRange;
+}
+
+/**
+ * Will walk up the scene object graph to the closest $editor scene object
+ */
+export function getSceneEditorFor(sceneObject: SceneObject): SceneEditor {
+  const { $editor } = sceneObject.state;
+  if ($editor) {
+    return $editor;
+  }
+
+  if (sceneObject.parent) {
+    return getSceneEditorFor(sceneObject.parent);
+  }
+
+  throw new Error('No editor found in scene tree');
+}
+
+/**
+ * Interpolates the given string using the current scene object as context.   *
+ */
+export function interpolateFor(sceneObject: SceneObject, value: string | undefined | null): string {
```

Comment:
I would probably change the order of arguments in the definition:

```suggestion
export function interpolateFor(value: string | undefined | null, sceneObject: SceneObject): string {
```

Interpolate `value` with `sceneObject` context.

Reply:
@dprokop I had it like that initially, the problem is that interpolate function will take more args (scopedVars, and custom format function), and maybe more args in the future. So given that signature it makes sense to have the context first, otherwise it would be come (target, sceneObject, arg2, arg3, etc )